File: src/test/java/io/github/ajoz/workshop/fp/part_1/solutions/exercise_4/Exercise4Test.java
package io.github.ajoz.workshop.fp.part_1.solutions.exercise_4;

import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.curry;
import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.flip;
import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.flipCurried;
import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.flipTupled;
import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.tuple;
import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.uncurry;
import static io.github.ajoz.workshop.fp.part_1.solutions.exercise_4.Exercise4.untuple;
import kotlin.Pair;
import static org.junit.Assert.assertEquals;
import org.junit.Test;

public class Exercise4Test {
    @Test
    public void should_convert_to_Function1_with_a_tuple() {
        // given:
        final Function2<Integer, Integer, Integer> add = (a, b) -> a + b;
        // when:
        final Function1<Pair<Integer, Integer>, Integer> tested = tuple(add);
        // then:
        assertEquals(add.apply(1, 1), tested.apply(new Pair<>(1, 1)));
        assertEquals(add.apply(0, 0), tested.apply(new Pair<>(0, 0)));
        assertEquals(add.apply(-1, 1), tested.apply(new Pair<>(-1, 1)));
    }

    @Test
    public void should_convert_to_Function1_with_a_Function1_return_type() {
        // given:
        final Function2<Integer, Integer, Integer> add = (a, b) -> a + b;
        // when:
        final Function1<Integer, Function1<Integer, Integer>> tested = curry(add);
        // then:
        assertEquals(add.apply(1, 1), tested.apply(1).apply(1));
        assertEquals(add.apply(0, 0), tested.apply(0).apply(0));
        assertEquals(add.apply(-1, 1), tested.apply(-1).apply(1));
    }

    @Test
    public void should_convert_to_Function2_from_a_Function1_with_a_tuple() {
        // given:
        final Function1<Pair<Integer, Integer>, Integer> add =
                ab_pair -> ab_pair.getFirst() + ab_pair.getSecond();
        // when:
        final Function2<Integer, Integer, Integer> tested = untuple(add);
        // then:
        assertEquals(add.apply(new Pair<>(1, 1)), tested.apply(1, 1));
        assertEquals(add.apply(new Pair<>(0, 0)), tested.apply(0, 0));
        assertEquals(add.apply(new Pair<>(-1, 1)), tested.apply(-1, 1));
    }

    @Test
    public void should_convert_to_Function2_from_a_Function1_with_a_Function1_as_return_type() {
        // given:
        final Function1<Integer, Function1<Integer, Integer>> add =
                a -> b -> a + b;
        // when:
        final Function2<Integer, Integer, Integer> tested =
                uncurry(add);
        // then:
        assertEquals(add.apply(1).apply(1), tested.apply(1, 1));
        assertEquals(add.apply(0).apply(0), tested.apply(0, 0));
        assertEquals(add.apply(-1).apply(1), tested.apply(-1, 1));
    }

    @Test
    public void should_flip_Function2() {
        // given:
        final Function2<String, Integer, String> function =
                (string, integer) -> String.format("%s : %d", string, integer);

        // when:
        final Function2<Integer, String, String> tested = flip(function);

        // then:
        assertEquals(function.apply("foo", 42), tested.apply(42, "foo"));
    }

    @Test
    public void should_flip_Function1_with_a_Tuple_argument() {
        // given:
        final Function1<Pair<String, Integer>, String> function =
                pair -> String.format("%s : %d", pair.getFirst(), pair.getSecond());

        // when:
        final Function1<Pair<Integer, String>, String> tested = flipTupled(function);

        // then:
        assertEquals(function.apply(new Pair<>("foo", 42)), tested.apply(new Pair<>(42, "foo")));
    }

    @Test
    public void should_flip_curried_function() {
        // given:
        final Function1<String, Function1<Integer, String>> function =
                string -> integer -> String.format("%s : %d", string, integer);

        // when:
        final Function1<Integer, Function1<String, String>> tested = flipCurried(function);

        // then:
        assertEquals(function.apply("foo").apply(42), tested.apply(42).apply("foo"));
    }
}

File: kotlin/exercises/src/test/kotlin/io/github/ajoz/workshop/fp/part1/exercise4/Exercise4Test.kt
package io.github.ajoz.workshop.fp.part1.exercise4

import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.experimental.runners.Enclosed
import org.junit.runner.RunWith

@RunWith(Enclosed::class)
class Exercise4Test {
    class Tuple {
        @Test
        fun `should convert to a one argument function`() {
            // given:
            val add: (Int, Int) -> Int = { a, b -> a + b }
            // when:
            val tested = tuple(add)
            // then:
            assertEquals(add(1, 1), tested(Pair(1, 1)))
            assertEquals(add(0, 0), tested(Pair(0, 0)))
            assertEquals(add(-1, 1), tested(Pair(-1, 1)))
        }

        @Test
        fun `should convert to a two argument function`() {
            // given:
            val add: (Pair<Int, Int>) -> Int = { ab_pair -> ab_pair.first + ab_pair.second }
            // when:
            val tested = unTuple(add)
            // then:
            assertEquals(add(Pair(1, 1)), tested(1, 1))
            assertEquals(add(Pair(0, 0)), tested(0, 0))
            assertEquals(add(Pair(-1, 1)), tested(-1, 1))
        }
    }

    class Curry {
        @Test
        fun `should convert a two argument function`() {
            // given:
            val add: (Int, Int) -> Int = { a, b -> a + b }
            // when:
            val tested = curry(add)
            // then:
            assertEquals(add(1, 1), tested(1)(1))
            assertEquals(add(0, 0), tested(0)(0))
            assertEquals(add(-1, 1), tested(-1)(1))
        }

        @Test
        fun `should convert back to a two argument function`() {
            // given:
            val add: (Int) -> (Int) -> Int = { a -> { b -> a + b } }
            // when:
            val tested = unCurry(add)
            // then:
            assertEquals(add(1)(1), tested(1, 1))
            assertEquals(add(0)(0), tested(0, 0))
            assertEquals(add(-1)(1), tested(-1, 1))
        }
    }

    class Flip {
        @Test
        fun `should flip a two argument function`() {
            // given:
            val function: (String, Int) -> String = { string, integer ->
                String.format("%s : %d", string, integer)
            }

            // when:
            val tested = flip(function)

            // then:
            assertEquals(function("foo", 42), tested(42, "foo"))
        }

        @Test
        fun `should flip a one argument function with a Tuple argument`() {
            // given:
            val function: (Pair<String, Int>) -> String = { pair ->
                String.format("%s : %d", pair.first, pair.second)
            }

            // when:
            val tested = flipTupled(function)

            // then:
            assertEquals(function(Pair("foo", 42)), tested(Pair(42, "foo")))
        }

        @Test
        fun `should flip a curried function`() {
            // given:
            val function: (String) -> (Int) -> String = { string ->
                { integer ->
                    String.format("%s : %d", string, integer)
                }
            }

            // when:
            val tested = flipCurried(function)

            // then:
            assertEquals(function("foo")(42), tested(42)("foo"))
        }
    }
}
